platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/ConfigurationCacheAwareBuildTreeWorkController.kt

```diff
@@ -41,22 +41,32 @@ class ConfigurationCacheAwareBuildTreeWorkController(
                 addFinalization(rootBuildState, selector::postProcessExecutionPlan)
             }
         }
-        val executionResult = workGraph.withNewWorkGraph { graph ->
+        val executionResult: ExecutionResult<Void>? = workGraph.withNewWorkGraph { graph ->
             val result = cache.loadOrScheduleRequestedTasks(
                 graph = graph,
                 graphBuilder = scheduleTaskSelectorPostProcessing
             ) { workPreparer.scheduleRequestedTasks(graph, taskSelector) }
-            if (!result.wasLoadedFromCache && !result.entryDiscarded && startParameter.loadAfterStore) {
-                // Load the work graph from cache instead
+            // There are four outcomes:
+            // 1. CC miss, graph has been successfully stored. We don't try to execute the graph directly but store it first, discard, and then reload.
+            // 2. Same as (1) but we also need tooling models. The model builders can be executed after the tasks (if any) in a build action,
+            //    and these builders may access project state as well as the task state. Because of that we execute the prepared graph directly.
+            // 3. CC miss, graph has been configured but the cached state discarded without failing the build (e.g. task.notCompatibleWithCC is used).
+            //    We execute the build immediately using the prepared graph.
+            // 4. CC hit: we've loaded the cached graph. We execute the build immediately using the loaded graph.
+            if (!result.wasLoadedFromCache && !result.entryDiscarded && !startParameter.isRequiresToolingModels) {
```

Comment:
💅  the condition will be much simpler to read if we swapped the if-else branches.

Reply:
> 💅 the condition will be much simpler to read if we swapped the if-else branches.

Perhaps, but this way it matches the order of the comment above better (I tried to change the comment but it reads better that way). Also, having a common case in the `then` branch is something that "Code Complete" recommends.